### PR TITLE
ML_E_step

### PR DESCRIPTION
**Pull request by VITYANA**

### CODE REVIEW

File: mpest/em/methods/likelihood_method.py [hunk 1]
```diff
@@ -59,17 +59,127 @@ def step(self, problem: Problem) -> EResult:
...
-#
-#         :param problem: Object of class Problem, which contains samples and mixture.
-#         """
+import numpy as np
+from typing import List
+from sklearn.cluster import KMeans, DBSCAN, AgglomerativeClustering
+from scipy.stats import weibull_min
+
+
+class ML(AExpectation[EResult]):
```
Comment (iraedeus) [inline mpest/em/methods/likelihood_method.py:68]: The order of actions in this class is wrong. First you initialize the mixture and evaluate it with the clusterizer, then save it inside the class. Then you put a mark that everything is initialized and at each beginning of the E step you pull this initialized mixture, which does not change in any way, you do not write there the mixture that comes from the M step after the iteration.

File: mpest/em/methods/likelihood_method.py [hunk 2]
```diff
@@ -59,17 +59,127 @@ def step(self, problem: Problem) -> EResult:
...
+            agglo = AgglomerativeClustering(n_clusters=self._n_components)
+            labels = agglo.fit_predict(X.reshape(-1, 1))
+        else:
+            raise ValueError("Can't find this clustering method.")
+
+        distributions = []
+        weights = []
+
+        for k in range(self._n_components):
+            X_k = X[labels == k]
```
Comment (iraedeus) [inline mpest/em/methods/likelihood_method.py:112]: What if labels correspond to first component, but label value is 1?

File: mpest/em/methods/likelihood_method.py [hunk 3]
```diff
@@ -59,17 +59,127 @@ def step(self, problem: Problem) -> EResult:
...
+            distributions.append(
+                DistributionInMixture(
+                    model=model,
+                    params=params,
+                    prior_probability=weight
+                )
+            )
+            weights.append(weight)
+
+        self._current_mixture = MixtureDistribution.from_distributions(distributions, weights)
```
[the remark this answers is not in the document]
Comment (VITYANA) [reply mpest/em/methods/likelihood_method.py]: fixed

File: mpest/em/methods/likelihood_method.py [hunk 4]
```diff
@@ -80,15 +77,15 @@ class ML(AExpectation[EResult]):
...
-    def __init__(self,models: list[AModel],method: str="kmeans",eps: float=0.3,accurate_init: bool=False) -> None:
+    def __init__(self,models: list[AModel],labels ,eps: float=0.3,accurate_init: bool=False) -> None:
         self._n_components = len(models)
-        self._method = method
         self._models = models
         self._initialized = False
         self._current_mixture = MixtureDistribution([])
         self._eps = eps
         self._accurate_init_flag = accurate_init
+        self._labels = labels
```
Comment (iraedeus) [inline mpest/em/methods/likelihood_method.py:87]: I meant moving the clusterer initialization outside the ML class, not moving and counting labels outside of it.

File: mpest/em/methods/likelihood_method.py [hunk 5]
```diff
@@ -99,10 +96,10 @@ def _estimate_weibull_params(data: np.ndarray) -> list[float]:
         except (ValueError, TypeError, FitError):
             return [0.5, np.mean(data)]
 
-    def _find_best_cluster_for_model(self, clusters, model):
+    def _find_best_cluster_for_model(self, clusters, model, min_samples=2):
```
Comment (iraedeus) [inline mpest/em/methods/likelihood_method.py:99]: I meant that it is better to put constants inside the class, but not inside the methods. Just do:
```python
Class ML:
    MIN_SAMPLES=2
    MIN_PROB=1e-100
```

**Review by iraedeus — CHANGES_REQUESTED** [review]
